### PR TITLE
[WIP] Refactor webchat for multiple providers

### PR DESCRIPTION
**Pull request by RobBond-GDS**
Refactor the existing webchat functionality provided by the Contacts App to allow for additional web chat providers.

Historically Contacts App was used to display webchat on HMRC only contact pages using one specific webchat provider (Egain). 

Refactoring has been done to create provider classes which contian there own specific configuration and logic, for example, the page webchat should be displayed on and provider availability endpoints

The need for this was to allow HM Passport Office to make use of the webchat and Contacts App functionality, as well as any future departments.

https://govuk.zendesk.com/agent/tickets/3540368

---

Visual regression results:
https://government-frontend-pr-1586.surge.sh/gallery.html

Component guide for this PR:
https://government-frontend-pr-1586.herokuapp.com/component-guide

### CODE REVIEW

File: app/assets/javascripts/webchat/providers/Klick2Contact.js [hunk 1]
```diff
@@ -3,7 +3,7 @@ function Klick2Contact(options) {
   this.k2c_provider             = 'HMPO2'; // Provider Name HMPO2 or HMPO
   this.k2c_url                  = 'https://hmpowebchat.klick2contact.com/v03'; // Url to console.
   this.k2c_launchServe          = 'https://hmpowebchat.klick2contact.com/v03'; // '', 'd' 's'
-  this.k2c_staticDept           = "712" // 712 or 717
+  this.k2c_staticDept           = "717" // 712 or 717
```
Comment (DilwoarH) [inline app/assets/javascripts/webchat/providers/Klick2Contact.js]: Could we pass this information via the config from the backend?

**Comment by DilwoarH** [comment]
Tests failed:

```
+ bundle exec rake default
Running `"phantomjs"  "/var/lib/jenkins/bundles/ruby/2.6.0/gems/jasmine-rails-0.15.0/lib/jasmine_rails/../assets/javascripts/jasmine-runner.js" "file:///var/lib/jenkins/workspace/r-webchat-for-multiple-providers/tmp/jasmine/runner.html?root=/var/lib/jenkins/workspace/r-webchat-for-multiple-providers&spec="`
Running: file:///var/lib/jenkins/workspace/r-webchat-for-multiple-providers/tmp/jasmine/runner.html?root=/var/lib/jenkins/workspace/r-webchat-for-multiple-providers&spec=
Starting...
[31mKlick2Contact should return available if agent is available[0m
[31m  Expected 'UNAVAILABLE' to be 'AVAILABLE'.
    error properties: Object({ sourceId: 140352268103936, stackArray: [ Object({ function: 'stack', sourceURL: 'file:///var/lib/jenkins/workspace/r-webchat-for-multiple-providers/tmp/jasmine/government-frontend/jasmine.js', line: 3586 }), Object({ function: 'buildExpectationResult', sourceURL: 'file:///var/lib/jenkins/workspace/r-webchat-for-multiple-providers/tmp/jasmine/government-frontend/jasmine.js', line: 3551 }), Object({ sourceURL: 'file:///var/lib/jenkins/workspace/r-webchat-for-multiple-providers/tmp/jasmine/government-frontend/jasmine.js', line: 1318 }), Object({ sourceURL: 'file:///var/lib/jenkins/workspace/r-webchat-for-multiple-providers/tmp/jasmine/government-frontend/jasmine.js', line: 701 }), Object({ function: 'addExpectationResult', sourceURL: 'file:///var/lib/jenkins/workspace/r-webchat-for-multiple-providers/tmp/jasmine/government-frontend/jasmine.js', line: 1232 }), Object({ sourceURL: 'file:///var/lib/jenkins/workspace/r-webchat-for-multiple-providers/tmp/jasmine/government-fronten ...
Error: Expected 'UNAVAILABLE' to be 'AVAILABLE'.
    at <Jasmine>[0m

Finished
-----------------
[31m28 specs, 1 failure in 0.172s.[0m

```

**Comment by leenagupte** [comment]
The tests are still failing:

```
NoMethodError:         NoMethodError: undefined method `webchat_availability_url' for #<ContactPresenter:0x000055753fa08e80>
           ​test/presenters/contact_presenter_test.rb:91:in `block in <class:PresentedContact>'
```

File: spec/javascripts/webchat-providers/klick-2-contact.spec.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+    expect(klick2contact.apiResponseSuccess(result).status).toBe("UNAVAILABLE")
+  })
+
+  it('should return error for when the API responds with an error', function () {
+    expect(klick2contact.apiResponseError({}).status).toBe("ERROR")
+  })
+
+  it('should return a valid open url', function() {
+    expect(klick2contact.openUrl()).toMatch(/([a-z]{1,2}tps?):\/\/((?:(?!(?:\/|#|\?|&)).)+)(?:(\/(?:(?:(?:(?!(?:#|\?|&)).)+\/))?))?(?:((?:(?!(?:\.|$|\?|#)).)+))?(?:(\.(?:(?!(?:\?|$|#)).)+))?(?:(\?(?:(?!(?:$|#)).)+))?(?:(#.+))?/)
```
Comment (leenagupte) [inline spec/javascripts/webchat-providers/klick-2-contact.spec.js]: Is there a human-readable URL that can be used here?

**Comment by RobBond-GDS** [comment]
Hi,

Be good to have conversations as soon as you are available, bit of pressure
to get this in as soon as possible as HM Passport office has been on the
cards for some time

when would be good for you?

thanks,

Rob

PS. have updated the test test that was failing....was still looking to
test functions that aren't there anymore!

**Comment by leenagupte** [comment]
> Hi, Be good to have conversations as soon as you are available, bit of pressure to get this in as soon as possible as HM Passport office has been on the cards for some time when would be good for you? thanks, Rob PS. have updated the test test that was failing....was still looking to test functions that aren't there anymore!

Can you put something in mine and @DilwoarH's calendars?

**Comment by DilwoarH** [comment]
Superseded by https://github.com/alphagov/government-frontend/pull/1594

**Pull request closed without merging**